azurecosmos/src/main/java/site/ycsb/db/AzureCosmosClient.java

```diff
@@ -187,218 +245,251 @@ private int getIntProperty(String propertyName, int defaultValue) {
+        List<ObjectNode> pageDocs = pageIterator.next().getResults();
+        for (ObjectNode doc : pageDocs) {
+          Map<String, String> stringResults = new HashMap<>();
+          Iterator<Map.Entry<String, JsonNode>> nodeIterator = doc.fields();
+          while (nodeIterator.hasNext()) {
+            Entry<String, JsonNode> pair = nodeIterator.next();
+            stringResults.put(pair.getKey().toString(), pair.getValue().toString());
+          }
+          HashMap<String, ByteIterator> byteResults = new HashMap<>();
+          StringByteIterator.putAllAsByteIterators(byteResults, stringResults);
+          result.add(byteResults);
+        }
       }
-      LOGGER.error("Failed to scan with startKey={}, recordCount={}", startkey, recordcount, e);
-      return Status.ERROR;
-    }
-
-    if (documents != null) {
-      for (Document document : documents) {
-        result.add(this.extractResult(document));
+      return Status.OK;
+    } catch (CosmosException e) {
+      if (!AzureCosmosClient.includeExceptionStackInLog) {
+        e = null;
       }
+      LOGGER.error("Failed to query key {} from collection {} in database {}", startkey, table,
+          AzureCosmosClient.databaseName, e);
     }
-
-    return Status.OK;
+    return Status.ERROR;
   }
 
+  /**
+   * Update a record in the database. Any field/value pairs in the specified
+   * values HashMap will be written into the record with the specified record key,
+   * overwriting any existing values with the same field name.
+   *
+   * @param table  The name of the table
+   * @param key    The record key of the record to write.
+   * @param values A HashMap of field/value pairs to update in the record
+   * @return Zero on success, a non-zero error code on error
+   */
   @Override
   public Status update(String table, String key, Map<String, ByteIterator> values) {
-    // Azure Cosmos does not have patch support.  Until then we need to read
-    // the document, update in place, and then write back.
-    // This could actually be made more efficient by using a stored procedure
-    // and doing the read/modify write on the server side.  Perhaps
-    // that will be a future improvement.
-
-    String documentLink = getDocumentLink(this.databaseName, table, key);
-    ResourceResponse<Document> updatedResource = null;
-    ResourceResponse<Document> readResouce = null;
-    RequestOptions reqOptions = null;
-    Document document = null;
 
-    try {
-      reqOptions = getRequestOptions(key);
-      readResouce = AzureCosmosClient.client.readDocument(documentLink, reqOptions);
-      document = readResouce.getResource();
-    } catch (DocumentClientException e) {
-      if (!this.includeExceptionStackInLog) {
-        e = null;
-      }
-      LOGGER.error("Failed to read key {} in collection {} in database {} during update operation",
-          key, table, this.databaseName, e);
-      return Status.ERROR;
-    }
+    int numAttempts = 4;
```

Comment:
should this be a YCSB config?

Reply:
My bad :) probably makes sense for Armaan to add this since he is doing the PR.

Reply:
Andy added this so I didn't touch it. Is a retry wrapper around update necessary? If so, I'll make "numAttempts" a constant in the file.

Reply:
Is a retry wrapper around update necessary? If so, I'll make "numAttempts" a constant in the file.

Reply:
Hi Armaan. Sorry just saw this comment. At this time it seems we may want to keep update retries. The reason is that multiple threads in YCSB may concurrently modify the same document, which may lead to service-side lock contention causing one or more concurrent modifications to be rejected (449 error). 

A config setting in azurecosmos.properties for update-retry-count would allow the user to make a latency/success-rate tradeoff for updates. 

The current retry-count of 4 was chosen arbitrarily so that we could test the feature.